File: leetcode77.cpp
class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<vector<int> > res;
        vector<int> tmp(k,0);
        int i = 0;
        while(i >= 0)
        {
        	tmp[i]++;
        	if(tmp[i] > n)
        		i--;
        	else if(i == k-1)
        		res.push_back(tmp);
        	else
        	{
        		i++;
        		tmp[i] = tmp[i-1];
        	}
        }
        return res;
    }
};

File: leetcode77.java
public class Solution {
    public List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> res = new ArrayList<>();
        backtrack(res, new ArrayList<Integer>(), 1, n, k);
        return res;
    }
    
    private void backtrack(List<List<Integer>> res, List<Integer> cur,int start, int n, int k){
        // end condition
        if(k == 0){
            res.add(new ArrayList<Integer>(cur));
            return;
        }
        for(int i = start; i <= n; i++){
            cur.add(i);
            backtrack(res,cur,i+1,n,k-1);
            cur.remove(cur.size()-1);
        }
    }
}
